enhancements/microshift/microshift-apiserver-custom-certs.md

```diff
@@ -0,0 +1,223 @@
+
+### User Stories
+* As a MicroShift administrator, I want to be able to add  organization generated certificates. 
+  - each certificate may contain:
+    - Single Common Name containing The API server DNS/IPAddress or a wildcard entry (wildcard certificate).
+    - Multiple Subject Alternative Names (SAN) containing the API server DNS/IPAddress.
+
+* As a MicroShift administrator, I want to provide additional DNS names for each certificate (in the Microshift config).
+
+### Goals
+* Allow MicroShift administrator to configure Microshift with externally generated certificates and
+domain names.
+
+### Non-Goals
+Automatic certificate rotation and integration with 3rd party cert issuing services.
+
+## Proposal
+Proposal suggests to provide Administrator means adding their own Certificates using microshift configuration file (/etc/microshift/config.yaml).
+
+A new `apiServer` level section will be added to the configuration called `namedCertificates`:
+
+```yaml
+apiServer:
+  namedCertificates:
+  - certPath: ~/certs/api_fqdn_1.crt
+    keyPath:  ~/certs/api_fqdn_1.key
+  - certPath: ~/certs/api_fqdn_2.crt
+    keyPath:  ~/certs/api_fqdn_2.key
+    names:
+    - api_fqdn_1     
+    - *.apps.external.com
+
+```  
+For each provided certificate, the following configuration is proposed:
+1. `names` - optional list of explicit DNS names (leading wildcards allowed).
+   If no names are provided, the implicit names will be extracted from the certificates.
+1. `certPath` -  certificate full path.
+1. `keyPath` -   certificate key full path.
+
+Certificate files will be read from their configured location by the Microshift service,
+each certification file will be validated (see validation rules).
+
+Because we dont want to disrupt the internal API communication and make sure the internal certificate will be automaticly renewed,
+Those certificate will extend the default  API server [external](https://github.com/openshift/microshift/blob/main/pkg/controllers/kube-apiserver.go#L194) certificate configuration.
+
+Certificate will be validated and treated according to [Failure Modes](#failure-modes).
+
+### Kubeconfig Generation
+Each configured certificate can contain multiple FQDN and wildcards values, for each unique FQDN address kubeconfig file will be generated on the filesystem.
+
+Every generated `kubeconfig` for the custom certificates will omit the certificate-authority-data section,
+therefore custom certificates will have to be validated against CAs in the RHEL Client trust store. 
+
+each certificate will be examined and deteremined if its wildcard only.
+certificate will be considered as a wildcard only if there are no FQDN Entries found.
+the FQDN Address is searched in:
+- Certificate CN
+- Certificate Subject Alternative Name (SAN)
+- names configuration value.
+
+when a Certificate is found to be a wildcard only, Microshift will not be able to find the real FQDN, node IP Address will be placed inside its kubeconfig's `server` section.
```

Comment:
@stlaz you and @eslutsky are using different terminology. Are you talking about the same thing? Are you saying that if a cert _only_ has wildcard names then it's not valid?

Reply:
@dhellmann ,  
according to the [rfc](https://www.ietf.org/rfc/rfc2818.txt) in context of `wildcard`
>  If a subjectAltName extension of type dNSName is present, that MUST
>    be used as the identity. Otherwise, the (most specific) Common Name
>    field in the Subject field of the certificate MUST be used. Although
>    the use of the Common Name is existing practice, it is deprecated and
>    Certification Authorities are encouraged to use the dNSName instead.
> 
>    Matching is performed using the matching rules specified by
>    [RFC2459].  If more than one identity of a given type is present in
>    the certificate (e.g., more than one dNSName name, a match in any one
>    of the set is considered acceptable.) Names may contain the wildcard
>    character * which is considered to match any single domain name
>    component or component fragment. E.g., *.a.com matches foo.a.com but
>    not bar.foo.a.com. f*.com matches foo.com but not bar.com.
> 

so  like @stlaz says , we should avoid using CN  since its deprecated ,
so we should only using the DNS  which is part  of the SAN.

Reply:
OK, what wasn't clear to me is you were talking about different fields in the certificate. If it's important that we treat the fields differently, please put that detail in the body of the enhancement.